driver/ppm_fillers.c

```diff
@@ -1184,6 +1186,9 @@ static int ppm_get_tty(void)
 
 	tty_nr = new_encode_dev(MKDEV(major, minor_start) + index);
 
+	if (snprintf(buf, buflen, "/dev/%s/%d", driver->name, index) < 0)
```

Comment:
Is this always right? What about all those cases written in https://github.com/draios/sysdig/issues/918#issuecomment-339601480 ?

Reply:
I have checked all the cases and got the same output as `tty` process

Reply:
On my machine I get `/dev/tty1` and that cannot be generated with this fmt string.

Reply:
I have tested it on Ubuntu/RHEL and in both I get the name as above `/dev/pts/6`.
On my mac OSX I get this `/dev/ttys001` but mac is not supported.

@gianlucaborello said that the full path is preferred but I can change it to the short version

Reply:
@arossert Are you connecting to them through SSH or similar? I get `/dev/pts/X` in that case, but if you go to the real terminal by using CTRL-ALT-F1 and do `tty` you get `/dev/tty1`

Reply:
I'm running the terminal from the GUI, I can try doing this that way

Reply:
I have looked again at this function from my comment #918 :
```static int guess_name(char *const buf, int maj, int min)```
and it seems that there are a lot of cases that can determine the path of the `tty` device.

I couldn't find anything in the kernel stracts that can give this information directly (maybe you can take a look)
If we want to go with the `guess_name()` we could put this code in the kernel or in the userspace.

Reply:
Oh, I see. What does driver->name() return? I thought it was the whole path. Is it simply the path suffix after /dev? You can compare it to the values you find in `/sys/dev/{block,char}/{major}:{minor}/uevent`.

I don't think you want to hardcode major numbers and make assumptions about what devices they represent. I'm sure they're fairly stable but I bet it's not universal across all linux distros.

Reply:
I couldn't find anything that holds the full path of the device.
What I could find is this:
Let take this device as example `dev/pts/12`

`struct tty_driver` has `const char *name;` member that represents `pts`
`struct tty_struct` has `char name[64]` member that represents `pts12`
`struct tty_struct` has `int index` member that represents `12`

This are the structs that I looked into, let me know if you can find something in here.
```
struct tty_struct {
	int	magic;
	struct kref kref;
	struct device *dev;
	struct tty_driver *driver;
	const struct tty_operations *ops;
	int index;

	/* Protects ldisc changes: Lock tty not pty */
	struct ld_semaphore ldisc_sem;
	struct tty_ldisc *ldisc;

	struct mutex atomic_write_lock;
	struct mutex legacy_mutex;
	struct mutex throttle_mutex;
	struct rw_semaphore termios_rwsem;
	struct mutex winsize_mutex;
	spinlock_t ctrl_lock;
	spinlock_t flow_lock;
	/* Termios values are protected by the termios rwsem */
	struct ktermios termios, termios_locked;
	struct termiox *termiox;	/* May be NULL for unsupported */
	char name[64];
	struct pid *pgrp;		/* Protected by ctrl lock */
	struct pid *session;
	unsigned long flags;
	int count;
	struct winsize winsize;		/* winsize_mutex */
	unsigned long stopped:1,	/* flow_lock */
		      flow_stopped:1,
		      unused:BITS_PER_LONG - 2;
	int hw_stopped;
	unsigned long ctrl_status:8,	/* ctrl_lock */
		      packet:1,
		      unused_ctrl:BITS_PER_LONG - 9;
	unsigned int receive_room;	/* Bytes free for queue */
	int flow_change;

	struct tty_struct *link;
	struct fasync_struct *fasync;
	wait_queue_head_t write_wait;
	wait_queue_head_t read_wait;
	struct work_struct hangup_work;
	void *disc_data;
	void *driver_data;
	spinlock_t files_lock;		/* protects tty_files list */
	struct list_head tty_files;

#define N_TTY_BUF_SIZE 4096

	int closing;
	unsigned char *write_buf;
	int write_cnt;
	/* If the tty has a pending do_SAK, queue it here - akpm */
	struct work_struct SAK_work;
	struct tty_port *port;
} __randomize_layout;
```

```
struct tty_driver {
	int	magic;		/* magic number for this structure */
	struct kref kref;	/* Reference management */
	struct cdev **cdevs;
	struct module	*owner;
	const char	*driver_name;
	const char	*name;
	int	name_base;	/* offset of printed name */
	int	major;		/* major device number */
	int	minor_start;	/* start of minor device number */
	unsigned int	num;	/* number of devices allocated */
	short	type;		/* type of tty driver */
	short	subtype;	/* subtype of tty driver */
	struct ktermios init_termios; /* Initial termios */
	unsigned long	flags;		/* tty driver flags */
	struct proc_dir_entry *proc_entry; /* /proc fs entry */
	struct tty_driver *other; /* only used for the PTY driver */

	/*
	 * Pointer to the tty data structures
	 */
	struct tty_struct **ttys;
	struct tty_port **ports;
	struct ktermios **termios;
	void *driver_state;

	/*
	 * Driver methods
	 */

	const struct tty_operations *ops;
	struct list_head tty_drivers;
} __randomize_layout;
```

Reply:
Since you can't find a complete match in kernel space, I think I'm back to suggesting it be done in userspace. You can do the `/sys/dev/...` lookup in userspace to match the major/minor number and construct the full path from there.

Reply:
I'm not sure how we can lookup the `tty` name from `/sys/dev/`.
And if we will do it in the userspace, it will not appear in the syscall arguments, correct?

Reply:
What I was originally thinking is that from tty_nr you can get the device major and minor number. https://stackoverflow.com/questions/19598467/getting-minor-device-number-from-tty-nr-attribute-in-proc-pid-stat points out the macros to use.

Once you have the major/minor number, you can look at `/sys/dev/{block,char}/{major}:{minor}`, which is a symlink to some directory below `/sys/devices/...`. In that directory, the file `uevent` has a DEVNAME field, which is the path below `/dev` for the device.

For pseudo-ttys aka /dev/pts/N, it's a bit more complicated, at those devices don't appear below `/sys/dev`. For those, you could also look at the devices /dev/pts/NNN, get their device numbers, and try to match them to the major/minor device numbers encoded in tty_nr.

I see that if you combine the tty_driver.name and tty_driver.index you do get 'pts/NNN`, so maybe that's good enough after all. You would still have to work out the discrepancy between pseudo tty and actual tty devices.

Reply:
It looks very complicated task if we are willing to use the TTY name as it appears in the kernel (like `pts12`) I can fix that.

Did you take a look at how [procps](https://github.com/shichao-an/procps-ng-3.3.10/blob/master/proc/devname.c#L286) translate the `tty_nr` to the name?
Do you think we can adopt this code?

Reply:
Of the 4 methods procps tries, `driver_name` seems very lightweight and likely to work most of the time, with entries to map device major/minor to a full /dev/xxxNNN name. It seems to handle the most common terminal types. How about you adapt the approach of reading /proc/tty/drivers (but not the code directly, that code is a different license than sysdig's), in userspace, from the tty_nr ?